src/System.Slices/System/Buffers/OwnedMemory.cs

```diff
@@ -120,8 +122,8 @@ protected internal virtual DisposableReservation Reserve(ref ReadOnlyMemory<T> m
         #region Used by Memory<T>
         void IKnown.AddReference(long id)
         {
-            VerifyId(id);
             AddReference();
+            VerifyId(id);
```

Comment:
So you have in Dispose you have 
   invalidate ID, followed by check reference count is zero
and here you have 
   ensure non-zero reference count, followed by validate ID. 

By performing the checks/validate after the action that will cause the other operation to fail you guarantee that at least one of the operations will fail if they race.